File: datasets/C++/Deepseek/DataStructures/easy5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <algorithm>
using namespace std;

struct Node {
    int value;
    Node* left;
    Node* right;
    Node(int val) : value(val), left(nullptr), right(nullptr) {}
};

void printTopView(Node* root) {
    if (root == nullptr) return;
    
    map<int, int> topView; // key: horizontal distance, value: node value
    queue<pair<Node*, int>> q; // pair of node and its horizontal distance
    q.push({root, 0});
    
    while (!q.empty()) {
        auto current = q.front();
        q.pop();
        Node* node = current.first;
        int hd = current.second;
        
        if (topView.find(hd) == topView.end()) {
            topView[hd] = node->value;
        }
        
        if (node->left != nullptr) {
            q.push({node->left, hd - 1});
        }
        if (node->right != nullptr) {
            q.push({node->right, hd + 1});
        }
    }
    
    for (auto const& entry : topView) {
        cout << entry.second << " ";
    }
    cout << endl;
}

int main() {
    int n;
    cin >> n;
    if (n == 0) return 0;
    
    vector<int> values(n);
    for (int i = 0; i < n; ++i) {
        cin >> values[i];
    }
    
    int e;
    cin >> e;
    
    map<int, Node*> nodes;
    for (int val : values) {
        nodes[val] = new Node(val);
    }
    
    Node* root = nodes[values[0]];
    
    for (int i = 0; i < e; ++i) {
        int parentVal, childVal;
        char direction;
        cin >> parentVal >> childVal >> direction;
        Node* parent = nodes[parentVal];
        Node* child = nodes[childVal];
        if (direction == 'L') {
            parent->left = child;
        } else {
            parent->right = child;
        }
    }
    
    printTopView(root);
    
    // Clean up memory
    for (auto entry : nodes) {
        delete entry.second;
    }
    
    return 0;
}

File: datasets/Python/Deepseek/DataStructures/easy5.py
from collections import deque

class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def print_top_view(root):
    if root is None:
        return
    
    # Dictionary to store the first node at each horizontal distance
    top_view = {}
    # Queue for level order traversal: stores (node, horizontal distance)
    queue = deque()
    queue.append((root, 0))
    
    while queue:
        node, hd = queue.popleft()
        
        # If the HD is encountered for the first time, add to dictionary
        if hd not in top_view:
            top_view[hd] = node.value
        
        if node.left:
            queue.append((node.left, hd - 1))
        if node.right:
            queue.append((node.right, hd + 1))
    
    # Sort the dictionary by HD and print the values
    for hd in sorted(top_view.keys()):
        print(top_view[hd], end=" ")
    print()

def main():
    n = int(input())
    if n == 0:
        return
    values = list(map(int, input().split()))
    e = int(input())
    
    # Create nodes
    nodes = {value: Node(value) for value in values}
    root = nodes[values[0]]  # Assuming the first value is the root unless specified otherwise
    
    for _ in range(e):
        parent_val, child_val, direction = input().split()
        parent = nodes[int(parent_val)]
        child = nodes[int(child_val)]
        if direction == 'L':
            parent.left = child
        else:
            parent.right = child
    
    print_top_view(root)

if __name__ == "__main__":
    main()